lib/route.js

```diff
@@ -208,117 +228,132 @@ function buildRouting (options) {
-        // be available after all of the plugins and routes have been loaded.
-        fourOhFour.setContext(this, context)
+          // Must store the 404 Context in 'preReady' because it is only guaranteed to
+          // be available after all of the plugins and routes have been loaded.
+          fourOhFour.setContext(this, context)
 
-        if (opts.schema) {
-          context.schema = normalizeSchema(context.schema)
+          if (opts.schema) {
+            context.schema = normalizeSchema(context.schema)
 
-          const schemaController = this[kSchemaController]
-          if (!opts.validatorCompiler && (opts.schema.body || opts.schema.headers || opts.schema.querystring || opts.schema.params)) {
-            schemaController.setupValidator(this[kOptions])
-          }
-          try {
-            compileSchemasForValidation(context, opts.validatorCompiler || schemaController.validatorCompiler)
-          } catch (error) {
-            throw new FST_ERR_SCH_VALIDATION_BUILD(opts.method, url, error.message)
-          }
+            const schemaController = this[kSchemaController]
+            if (!opts.validatorCompiler && (opts.schema.body || opts.schema.headers || opts.schema.querystring || opts.schema.params)) {
+              schemaController.setupValidator(this[kOptions])
+            }
+            try {
+              compileSchemasForValidation(context, opts.validatorCompiler || schemaController.validatorCompiler)
+            } catch (error) {
+              throw new FST_ERR_SCH_VALIDATION_BUILD(opts.method, url, error.message)
+            }
 
-          if (opts.schema.response && !opts.serializerCompiler) {
-            schemaController.setupSerializer(this[kOptions])
-          }
-          try {
-            compileSchemasForSerialization(context, opts.serializerCompiler || schemaController.serializerCompiler)
-          } catch (error) {
-            throw new FST_ERR_SCH_SERIALIZATION_BUILD(opts.method, url, error.message)
+            if (opts.schema.response && !opts.serializerCompiler) {
+              schemaController.setupSerializer(this[kOptions])
+            }
+            try {
+              compileSchemasForSerialization(context, opts.serializerCompiler || schemaController.serializerCompiler)
+            } catch (error) {
+              throw new FST_ERR_SCH_SERIALIZATION_BUILD(opts.method, url, error.message)
+            }
           }
+        })
+
+        const { exposeHeadRoute } = opts
+        const hasRouteExposeHeadRouteFlag = exposeHeadRoute != null
+        const shouldExposeHead = hasRouteExposeHeadRouteFlag ? exposeHeadRoute : globalExposeHeadRoutes
+
+        if (shouldExposeHead && options.method === 'GET' && !headRouteExists) {
+          const onSendHandlers = parseHeadOnSendHandlers(opts.onSend)
+          prepareRoute.call(this, 'HEAD', path, { ...opts, onSend: onSendHandlers })
+        } else if (headRouteExists && exposeHeadRoute) {
+          warning.emit('FSTDEP007')
```

Comment:
The error stated on [this comment](https://github.com/fastify/fastify/pull/2954#issuecomment-819051784) was caused by the order of how the sibling `HEAD` route was being registered for each `GET` one.

As each `GET` route was creating the sibling `HEAD` one before registering a handler for the `after` hook, the `HEAD` route was registering first the hook handler, changing the order on how the validations should be done.

I moved the register of the sibling route on the `after` route hook. Not sure if is the best or should do it after registering the hook handler and not within. Looking forward to having feedback 🙂

Reply:
Good for me